zcash_client_sqlite/src/wallet/sapling.rs

```diff
@@ -1325,6 +1326,161 @@ pub(crate) mod tests {
+
+        let data_file = NamedTempFile::new().unwrap();
+        let mut db_data = WalletDb::for_path(data_file.path(), tests::network()).unwrap();
+        init_wallet_db(&mut db_data, None).unwrap();
+
+        // Add an account to the wallet
+        let seed = Secret::new([0u8; 32].to_vec());
+        let (_, usk) = db_data.create_account(&seed).unwrap();
+        let dfvk = usk.sapling().to_diversifiable_full_viewing_key();
+
+        let (_, usk2) = db_data.create_account(&seed).unwrap();
+        let dfvk2 = usk2.sapling().to_diversifiable_full_viewing_key();
+
+        // Add funds to the wallet in a single note
+        let value = Amount::from_u64(100000).unwrap();
+        let (cb, _) = fake_compact_block(
+            sapling_activation_height(),
+            BlockHash([0; 32]),
+            &dfvk,
+            AddressType::DefaultExternal,
+            value,
+            0,
+        );
+        insert_into_cache(&db_cache, &cb);
+        scan_cached_blocks(
+            &tests::network(),
+            &db_cache,
+            &mut db_data,
+            sapling_activation_height(),
+            1,
+        )
+        .unwrap();
+
+        // Verified balance matches total balance
+        let (_, anchor_height) = db_data
+            .get_target_and_anchor_heights(NonZeroU32::new(1).unwrap())
+            .unwrap()
+            .unwrap();
+        assert_eq!(
+            get_balance(&db_data.conn, AccountId::from(0)).unwrap(),
+            value
+        );
+        assert_eq!(
+            get_balance_at(&db_data.conn, AccountId::from(0), anchor_height).unwrap(),
+            value
+        );
+        assert_eq!(
+            get_balance(&db_data.conn, AccountId::from(1)).unwrap(),
+            Amount::from_u64(0).unwrap()
+        );
+
+        let amount_sent = Amount::from_u64(20000).unwrap();
+        let addr = dfvk.default_address().1;
+        let addr2 = dfvk2.default_address().1;
+        let req = TransactionRequest::new(vec![
+            // payment to an external recipient
+            Payment {
+                recipient_address: RecipientAddress::Shielded(addr2),
+                amount: amount_sent,
+                memo: None,
+                label: None,
+                message: None,
+                other_params: vec![],
+            },
+            // payment back to the originating wallet, simulating legacy change
+            Payment {
+                recipient_address: RecipientAddress::Shielded(addr),
+                amount: Amount::from_u64(30000).unwrap(),
+                memo: None,
+                label: None,
+                message: None,
+                other_params: vec![],
+            },
+        ])
+        .unwrap();
+
+        let fee_rule = FixedFeeRule::standard();
+        let input_selector = GreedyInputSelector::new(
+            fixed::SingleOutputChangeStrategy::new(fee_rule),
+            DustOutputPolicy::default(),
+        );
+
+        let txid = spend(
```

Comment:
This will send change back to the internal address; is that intended for this test?

Reply:
Oh, right, the self-send should be detected as legacy change.